queue.go

```diff
@@ -0,0 +1,423 @@
+package maddy
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/emersion/maddy/config"
+	"github.com/emersion/maddy/log"
+	"github.com/emersion/maddy/module"
+	"github.com/google/uuid"
+)
+
+// PartialError describes state of partially successfull message delivery.
+type PartialError struct {
+	// Recipients for which delivery was successful.
+	Successful []string
+	// Recipients for which delivery permanently failed.
+	Failed []string
+	// Recipients for which delivery tmeporarly failed.
+	TemporaryFailed []string
+
+	// Underlying error objects.
+	Errs map[string]error
+}
+
+func (pe PartialError) Error() string {
+	return fmt.Sprintf("delivery failed for some recipients: %v", pe.Errs)
+}
+
+type Queue struct {
+	name     string
+	location string
+	wheel    *TimeWheel
+
+	initialRetryTime time.Duration
+	retryTimeScale   float64
+	maxTries         int
+
+	Log    log.Logger
+	Target module.DeliveryTarget
+
+	workersWg sync.WaitGroup
+}
+
+type QueueMetadata struct {
+	UUID uuid.UUID
+
+	Ctx *module.DeliveryContext
+
+	Failed []string
+
+	// Amount of times delivery *already tried*.
+	TriesCount  int
+	LastAttempt time.Time
+}
+
+func NewQueue(_, instName string) (module.Module, error) {
+	return &Queue{
+		name:             instName,
+		initialRetryTime: 30 * time.Second,
+		retryTimeScale:   5,
+		Log:              log.Logger{Name: "queue"},
+	}, nil
+}
+
+func (q *Queue) Init(cfg *config.Map) error {
+	q.wheel = NewTimeWheel()
+	var workers int
+
+	cfg.Bool("debug", true, &q.Log.Debug)
+	cfg.Int("max_tries", false, false, 4, &q.maxTries)
+	cfg.Int("workers", false, false, 16, &workers)
+	cfg.String("location", false, false, filepath.Join(StateDirectory(cfg.Globals), q.name), &q.location)
+	cfg.Custom("target", false, true, nil, deliverDirective, &q.Target)
+	if _, err := cfg.Process(); err != nil {
+		return err
+	}
+
+	if !filepath.IsAbs(q.location) {
+		q.location = filepath.Join(StateDirectory(cfg.Globals), q.location)
+	}
+
+	// TODO: Check location write permissions.
+	if err := os.MkdirAll(q.location, os.ModePerm); err != nil {
+		return err
+	}
+	if err := q.readDiskQueue(); err != nil {
+		return err
+	}
+
+	q.Log.Debugf("delivery target: %T", q.Target)
+
+	for i := 0; i < workers; i++ {
+		q.workersWg.Add(1)
+		go q.worker()
+	}
+
+	return nil
+}
+
+func (q *Queue) Close() error {
+	q.wheel.Stop()
+	q.workersWg.Wait()
+	return nil
+}
+
+func (q *Queue) worker() {
+	for slot := range q.wheel.Dispatch() {
+		q.Log.Debugln("worker woke up for", slot.Value)
+
+		id := slot.Value.(uuid.UUID)
+		meta, body, err := q.openMessage(id)
+		if err != nil {
```

Comment:
`openMessage` doesn't close `body` in case of error

Reply:
ce1b726